"Conflicting definitions for ‘a’" error when desugaring data type

The following program:

```hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE GHC2021 #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE UndecidableInstances #-}
{-# OPTIONS_GHC -ddump-splices -dsuppress-uniques #-}
module Bug where

import Data.Singletons.TH.Options
import Data.Singletons.Base.TH

$(withOptions defaultOptions{genSingKindInsts = False} $
  singletons [d|
  data D a (x :: Maybe a) = MkD
  |])
```

Will cause `singletons-th` to emit ill-scoped code:

```
[1 of 1] Compiling Bug              ( Bug.hs, interpreted )
Bug.hs:(12,2)-(15,5): Splicing declarations
    withOptions defaultOptions {genSingKindInsts = False}
      $ singletons [d| data D a (x :: Maybe a) = MkD |]
  ======>
    data D a (x :: Maybe a) = MkD
    type MkDSym0 :: forall a a (x :: Maybe a). D a (x :: Maybe a)
    type family MkDSym0 :: D a (x :: Maybe a) where
      MkDSym0 = MkD
    data SD :: forall a (x :: Maybe a). D a x -> Type
      where
        SMkD :: forall a a (x :: Maybe a). SD (MkD :: D a (x :: Maybe a))
    type instance Sing @(D a x) = SD
    instance SingI MkD where
      sing = SMkD

Bug.hs:12:2: error:
    Conflicting definitions for ‘a’
    Bound at: Bug.hs:(12,2)-(15,5)
              Bug.hs:(12,2)-(15,5)
   |
12 | $(withOptions defaultOptions{genSingKindInsts = False} $
   |  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...

Bug.hs:12:2: error:
    Conflicting definitions for ‘a’
    Bound at: Bug.hs:(12,2)-(15,5)
              Bug.hs:(12,2)-(15,5)
   |
12 | $(withOptions defaultOptions{genSingKindInsts = False} $
   |  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
```

Note the `forall a a` in the type of `SMkD`.

This bug is similar to goldfirere/singletons#567, but it is caused by a different part of the code. (Note that you don't need a standalone kind signature to trigger this bug.)

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -5,6 +5,8 @@ Version next [????.??.??]
 -------------------------
 * Fix a bug in which infix data family declaration would mistakenly be rejected
   when reified locally.
+* Fix a bug in which data types that use visible dependent quantification would
+  produce ill-scoped code when desugared.
 
 Version 1.15 [2023.03.12]
 -------------------------

--- a/Language/Haskell/TH/Desugar.hs
+++ b/Language/Haskell/TH/Desugar.hs
@@ -96,7 +96,8 @@ module Language.Haskell.TH.Desugar (
   unboxedSumDegree_maybe, unboxedSumNameDegree_maybe,
   unboxedTupleDegree_maybe, unboxedTupleNameDegree_maybe,
   isTypeKindName, typeKindName, bindIP,
-  mkExtraDKindBinders, dTyVarBndrToDType, changeDTVFlags, toposortTyVarsOf,
+  mkExtraDKindBinders, dTyVarBndrToDType, changeDTVFlags,
+  toposortTyVarsOf, toposortKindVarsOfTvbs,
 
   -- ** 'FunArgs' and 'VisFunArg'
   FunArgs(..), ForallTelescope(..), VisFunArg(..),

--- a/Language/Haskell/TH/Desugar/Core.hs
+++ b/Language/Haskell/TH/Desugar/Core.hs
@@ -26,6 +26,8 @@ import Control.Monad.Zip
 import Data.Data (Data)
 import Data.Either (lefts)
 import Data.Foldable as F hiding (concat, notElem)
+import Data.Function (on)
+import qualified Data.List as L
 import qualified Data.Map as M
 import Data.Map (Map)
 import Data.Maybe (isJust, mapMaybe)
@@ -959,7 +961,7 @@ dsCon univ_dtvbs data_type con = do
             expl_dtvbs = changeDTVFlags SpecifiedSpec univ_dtvbs ++
                          ex_dtvbs
             impl_dtvbs = changeDTVFlags SpecifiedSpec $
-                         toposortTyVarsOf $ mapMaybe extractTvbKind expl_dtvbs in
+                         toposortKindVarsOfTvbs expl_dtvbs in
         DCon (impl_dtvbs ++ expl_dtvbs) dcxt n fields data_type
       Just gadt_type ->
         let univ_ex_dtvbs = dtvbs in
@@ -1678,6 +1680,25 @@ toposortTyVarsOf tys =
 
   in map ascribeWithKind $
      scopedSort freeVars
+
+-- | Take a telescope of 'DTyVarBndr's, find the free variables in their kinds,
+-- and sort them in reverse topological order to ensure that they are well
+-- scoped. Because the argument list is assumed to be telescoping, kind
+-- variables that are bound earlier in the list are not returned. For example,
+-- this:
+--
+-- @
+-- 'toposortKindVarsOfTvbs' [a :: k, b :: Proxy a]
+-- @
+--
+-- Will return @[k]@, not @[k, a]@, since @a@ is bound earlier by @a :: k@.
+toposortKindVarsOfTvbs :: [DTyVarBndr flag] -> [DTyVarBndrUnit]
+toposortKindVarsOfTvbs tvbs =
+  foldr (\tvb kvs ->
+          foldMap (\t -> toposortTyVarsOf [t]) (extractTvbKind tvb) `L.union`
+          L.deleteBy ((==) `on` dtvbName) tvb kvs)
+        []
+        (changeDTVFlags () tvbs)
 
 dtvbName :: DTyVarBndr flag -> Name
 dtvbName (DPlainTV n _)    = n

--- a/Test/Run.hs
+++ b/Test/Run.hs
@@ -22,6 +22,10 @@ rae@cs.brynmawr.edu
 {-# LANGUAGE QuantifiedConstraints #-}
 #endif
 
+#if __GLASGOW_HASKELL__ < 806
+{-# LANGUAGE TypeInType #-}
+#endif
+
 #if __GLASGOW_HASKELL__ >= 809
 {-# LANGUAGE StandaloneKindSignatures #-}
 #endif
@@ -600,6 +604,22 @@ test_t171 =
          Just (DVarI _ actual_ty _) <- dsReify getT1
          expected_ty `eqTHSplice` actual_ty)
 
+-- A regression test for #188, which ensures that it produces the correct answer
+-- for an unusual telescope like:
+--
+--   ... forall (a1 :: a2). forall (a2 :: a1). ...
+--
+-- Here, a2 is free in the kind of a1 (the first `forall`), but then the second
+-- `forall` binds another a2 that shadows what was already in scope. In this
+-- example, `toposortKindVarsOfTvbs [(a1 :: a2), (a2 :: a1)]` should return
+-- [a2].
+test_t188 :: Bool
+test_t188 =
+  let a1 = mkName "a1"
+      a2 = mkName "a2" in
+  toposortKindVarsOfTvbs [DKindedTV a1 () (DVarT a2), DKindedTV a2 () (DVarT a1)]
+    == [DPlainTV a2 ()]
+
 -- Unit tests for functions that compute free variables (e.g., fvDType)
 test_fvs :: [Bool]
 test_fvs =
@@ -822,6 +842,8 @@ main = hspec $ do
 #endif
 
     it "locally reifies GADT record selector types with explicit foralls correctly" $ test_t171
+
+    it "computes free kind variables correctly in a telescope that uses shadowing" $ test_t188
 
     -- Remove map pprints here after switch to th-orphans
     zipWithM (\t t' -> it ("can do Type->DType->Type of " ++ t) $ t == t')

--- a/Test/Splices.hs
+++ b/Test/Splices.hs
@@ -29,6 +29,10 @@ rae@cs.brynmawr.edu
 {-# LANGUAGE QuantifiedConstraints #-}
 #endif
 
+#if __GLASGOW_HASKELL__ < 806
+{-# LANGUAGE TypeInType #-}
+#endif
+
 #if __GLASGOW_HASKELL__ >= 807
 {-# LANGUAGE ImplicitParams #-}
 #endif
@@ -676,6 +680,9 @@ reifyDecs = [d|
   -- A regression test for #184
   data family x ^^^ y
   data instance x ^^^ y = R40 x y
+
+  -- A regression test for #188
+  data R41 a (x :: Maybe a) = R42
   |]
 
 reifyDecsNames :: [Name]
@@ -695,6 +702,7 @@ reifyDecsNames = map mkName
   , "R36", "R37", "R38", "R39"
 #endif
   , "R40"
+  , "R41", "R42"
   ]
 
 simplCaseTests :: [Q Exp]

